lib/pbench/agent/tool_meister.py

```diff
@@ -1003,92 +1043,126 @@ def _send_client_status(self, status):
-            )
-            self._send_client_status("internal-error")
-            return False
-        failures = 0
-        tool_cnt = 0
-        for name, tool_opts in sorted(self._tools.items()):
-            if name not in self.persistent_tool_names:
-                continue
-            tool_cnt += 1
-            tklass = self._tool_name_class_mappings[name]
+            tool_dir.mkdir()
+        except Exception as exc:
+            raise ToolException(
+                "Failed to create local tool directory, %s", tool_dir
+            ) from exc
+        return tool_dir
+
+    def _start_tools(self, tools_to_start: Dict[str, Tool], tool_dir: Path) -> Dict[str, Tool]:
+        """Invoke the .start() method for each Tool in the dictionary of tools
+        to start.
+
+        Arguments:
+
+            tools_to_start:  dictionary of Tool objects to be started
+
+        Returns a dictionary of all the tools successfully tarted.
+        """
+        started_tools = {}
+        for name, tool in sorted(tools_to_start.items()):
             try:
-                persistent_tool = tklass(
-                    name,
-                    tool_opts,
-                    pbench_install_dir=self.pbench_install_dir,
-                    tool_dir=_tool_dir,
-                    logger=self.logger,
-                )
-                persistent_tool.start()
+                tool.start(tool_dir)
             except Exception:
                 self.logger.exception(
-                    "Failed to init PersistentTool %s running in background", name
+                    "Failure starting tool %s running in background", name
                 )
-                failures += 1
             else:
-                self._persistent_tools[name] = persistent_tool
-                self.logger.debug("NAME: " + name + "  TOOL OPTS: " + tool_opts)
+                started_tools[name] = tool
+        return started_tools
+
+    def init_tools(self, data: Dict[str, str]) -> int:
+        """Setup all registered persistent tools which have data collectors.
+
+        The Tool Data Sink will be setting up the actual processes which
+        collect data from these tools.
+
+        Arguments:
 
+            data: a dictionary of the arguments sent to the Tool Meister
+
+        Returns 0 on success, # of failures otherwise.
+        """
+        try:
+            _tool_dir = self._create_tool_directory(data["directory"])
+        except Exception as exc:
+            self.logger.exception(
+                "Failed to create local tool directory for %s", data["directory"]
+            )
+            self._send_client_status("internal-error")
+            return 1
+
+        # Remember this persistent tmp tool directory so that we can delete it
+        # when requested.
+        self.directories[data["directory"]] = _tool_dir.parent
+
+        # Start all the persistent tools running.
+        started_tools = self._start_tools(self._persistent_tools)
```

Comment:
LGTM points out this does *not* look good to [it] ... `_start_tools`(line 1105) requires a second `tool_dir` parameter.

Reply:
Yes, I merge a change that was not ready for review, unfortunately.

Reply:
This should be resolved now.